Do not try to save corrupt OpenTelemetry spans

qoega opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

On bad alloc we can have corrupt span that will be added to log in OpenTelemetrySpanHolder destructor.

qoega committed:
Do not try to save corrupt OpenTelemetry spans

novikd commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -690,6 +690,7 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                     std::string class_name(demangle(typeid(*raw_interpreter_ptr).name()));
                     span = std::make_unique<OpenTelemetrySpanHolder>(class_name + "::execute()");
                 }
+                span->ready();
```
Here `span` can be `nullptr`. You need to do it inside of if.

novikd commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -311,7 +311,7 @@ static void onExceptionBeforeStart(const String & query_for_logging, ContextPtr
             span.attributes.push_back(Tuple{"clickhouse.tracestate", context->query_trace_context.tracestate});
         }
-
+        span.is_ready = true;
```
```suggestion
        span.ready();
```

qoega replied:
It is not a Holder unfortunately.

novikd commented on src/Interpreters/OpenTelemetrySpanLog.h:
```diff
@@ -16,6 +16,7 @@ struct OpenTelemetrySpan
     UInt64 start_time_us;
     UInt64 finish_time_us;
     Map attributes;
+    bool is_ready;
```
```suggestion
    bool is_ready = false;
```

qoega committed:
Update src/Interpreters/OpenTelemetrySpanLog.h

Co-authored-by: Dmitry Novik <d1mas1k4@yandex.ru>

qoega commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -385,7 +385,7 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
                 std::chrono::system_clock::now().time_since_epoch()).count();
         span.attributes.push_back(Tuple{"clickhouse.thread_id", toString(thread_id)});
-
+        span.is_ready = true;
```
It is not a Holder unfortunately.

tavplubix replied:
So this line is redundant, because `is_ready` flag is checked in Holder's dtor only.

qoega committed:
fix

tavplubix commented on src/Interpreters/OpenTelemetrySpanLog.h:
```diff
@@ -48,6 +49,7 @@ struct OpenTelemetrySpanHolder : public OpenTelemetrySpan
     void addAttribute(const std::string& name, const std::string& value);
     void addAttribute(const Exception & e);
     void addAttribute(std::exception_ptr e);
+    void ready() { this->is_ready = true; }
```
Consider renaming it to `setReady`

tavplubix commented on src/Interpreters/OpenTelemetrySpanLog.cpp:
```diff
@@ -101,6 +101,11 @@ OpenTelemetrySpanHolder::~OpenTelemetrySpanHolder()
 {
     try
     {
+        // If span is not ready we don't try to save it.
```
Consider adding assertion like this:
```
assert(is_ready || std::uncaught_exceptions());
```

qoega committed:
improve

qoega committed:
rename

FrankChen021 commented:
Hi @qoega Could you show me an example of such corrupt span? I don't understand the problem this PR wants to fix. But the from the changed code, I think it looks like a little bit strange for every span to call such method.

qoega commented:
Spans are not exception safe. If you get bad_alloc for example in addAttribute during to string conversion SpanHolder destructor is still called and span is places to log.

FrankChen021 commented:
> Spans are not exception safe. If you get bad_alloc for example in addAttribute during to string conversion SpanHolder destructor is still called and span is places to log.

@qoega  I have different views. Once span is created, it means the execution enters a scoped flow, and the span logs reflects the execution. If there's no such span( 'query' for example), we know ClickHouse does not enter the code block of 'query'. So ignoring the span log introduces confusion.

Let's consider an edge case, if exception is thrown from such method immediately after the root span is created, there won't be any span recorded which is very strange that people will ask why the ClickHouse does not record any spans even the trace context has been passed from parent application to it and a HTTP response has been returned to the client?

IMO, If we're worried about exception thrown from `addAttribute`, it's better to capture it and ignore it.

If we DO want to ignore such span, why not capture the exception thrown from `addAttribute` and then mark the span as NOT_READY? Let the developers call an extra `setReady` API explicitly after  all `addAttribute` method calls is not safe.

qoega commented:
>So ignoring the span log introduces confusion.

You actually just got SIGSEGV in the 22.6 release in that case. I think it was much more confusion for developers. This case is really rare.
I don't think you want to make explicit call to end span instead of ~OpenTelemetrySpanHolder.
It is also possible to make span constructor to get all the data in the beginning and consider it ready.

FrankChen021 commented on src/Interpreters/OpenTelemetrySpanLog.cpp:
```diff
@@ -101,6 +101,14 @@ OpenTelemetrySpanHolder::~OpenTelemetrySpanHolder()
+        if (!is_ready)
+        {
+            assert(std::uncaught_exceptions());
+            return;
```
I think this `return` should be put after line 110 in original file where the context on current thread is restored.

qoega commented:
> @qoega AFAICS there should not be corrupted spans after using `Map` for `OpenTelemetrySpan::attributes`, what did I miss?

That is why I have not yet merged it. I also think there is no sense in additional checks there. All other fields in struct are safe to save to log even if they will not be filled yet if exception will be thrown.


> And just to understand, before (#38965) the problem was that it uses two `Array`'s for `attributes` and value was added to one array but not into another, and later when it iterates over attributes it trying to access non initialized value, is this correct?

The problem was that 
1) we add span to log in destructor. It is called even if we got bad_alloc while building a span.
2) There was a code that made `std::string(number)` in addAttribute after key was already added to attribute_names.
3) Everything worked before native Map support and corrupt spans just had unequal arrays there.
4) When Map type support was added to this table initial span struct was not changes and this error was added.
5) When we switched from Streams to Processors we lost old spans in Streams.
6) When we added Processors tracing it just increased probability to trigger a bug added in 4)

azat commented:
>There was a code that made std::string(number) in addAttribute after key was already added to attribute_names.

Yep. So only adding attributes was not safe, it can be solved by adding a helper that accept `name` and `value`, and rollback the changes.
Also destructor can check that the sizes are equal.

Anyway AFAICS there is no point in this patch (or #38965), since everything is safe already?
Plus this patch make the code a little bit complex while not bringing anything.
